src/platforms/stlinkv3/platform.c

```diff
@@ -0,0 +1,233 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file provides the platform specific functions for the STLINK-V3
+ * implementation.
+ */
+
+#include "general.h"
+#include "usb.h"
+#include "aux_serial.h"
+#include "gdb_if.h"
+
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/cm3/scb.h>
+#include <libopencm3/cm3/scs.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/usart.h>
+#include <libopencm3/usb/usbd.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/spi.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/syscfg.h>
+
+uint16_t srst_pin;
+static uint32_t hw_version;
+
+#define SCB_CCR_IC_SHIFT               17U                       /*!< SCB CCR: Instruction cache enable bit Position */
+#define SCB_CCR_IC_MASK                (1UL << SCB_CCR_IC_SHIFT) /*!< SCB CCR: Instruction cache enable bit Mask */
+#define SCB_CCR_DC_SHIFT               16U                       /*!< SCB CCR: Cache enable bit Position */
+#define SCB_CCR_DC_MASK                (1UL << SCB_CCR_DC_SHIFT) /*!< SCB CCR: DC Mask */
+#define SCB_CCSIDR_NUMSETS_MASK        (0x7fffUL << SCB_CCSIDR_NUMSETS_SHIFT) /*!< SCB CCSIDR: NumSets Mask */
+#define SCB_CCSIDR_NUMSETS_SHIFT       13U                                    /*!< SCB CCSIDR: NumSets Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_SHIFT 3U                                     /*!< SCB CCSIDR: Associativity Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_MASK  (0x3ffUL << SCB_CCSIDR_ASSOCIATIVITY_SHIFT) /*!< SCB CCSIDR: Associativity Mask */
+#define CCSIDR_WAYS(x)                 (((x)&SCB_CCSIDR_ASSOCIATIVITY_MASK) >> SCB_CCSIDR_ASSOCIATIVITY_SHIFT)
+#define CCSIDR_SETS(x)                 (((x)&SCB_CCSIDR_NUMSETS_MASK) >> SCB_CCSIDR_NUMSETS_SHIFT)
+#define SCB_DCISW_SET_SHIFT            5U                               /*!< SCB DCISW: Set Position */
+#define SCB_DCISW_SET_MASK             (0x1ffUL << SCB_DCISW_SET_SHIFT) /*!< SCB DCISW: Set Mask */
+#define SCB_DCISW_WAY_SHIFT            30U                              /*!< SCB DCISW: Way Position */
+#define SCB_DCISW_WAY_MASK             (3UL << SCB_DCISW_WAY_SHIFT)     /*!< SCB DCISW: Way Mask */
+
+static void cm_dsb(void)
+{
+	__asm__ volatile("dsb 0xf" ::: "memory");
+}
+
+static void cm_isb(void)
+{
+	__asm__ volatile("isb 0xf" ::: "memory");
+}
+
+static void scb_enable_i_cache(void)
+{
+	cm_dsb();
+	cm_isb();
+	SCB_ICIALLU = 0UL; /* invalidate I-Cache */
+	cm_dsb();
+	cm_isb();
+	SCB_CCR |= (uint32_t)SCB_CCR_IC_MASK; /* enable I-Cache */
+	cm_dsb();
+	cm_isb();
+}
+
+static void scb_enable_d_cache(void)
+{
+	SCB_CCSELR = 0U; /*(0U << 1U) | 0U;*/ /* Level 1 data cache */
+	cm_dsb();
+
+	uint32_t ccsidr = SCB_CCSIDR;
+
+	uint32_t sets = (uint32_t)CCSIDR_SETS(ccsidr);
+	do {
```

Comment:
The more we read this, the more this looks like a backwards for loop - the `do {} while ()` pattern here is obscuring meaning and the condition in the while component is bug-prone - for example, what happens if sets is 0 to start with? (same for ways)

This reads like it should be something like:
```c
const uint32_t sets =  CCSIDR_SETS(ccsidr);
const uint32_t ways = CCSIDR_WAYS(ccsidr);
for (uint32_t set = 0; set <= sets; ++set) {
    for (uint32_t way = 0; way <= ways; ++way) {
        CSB_DCISW = ((set << SCB_DCISW_SET_SHIFT) & SCB_DCISW_SET_MASK) |
            ((way << SCB_DCISW_WAY_SHIFT) & SCB_DCISW_WAY_MASK);
    }
}
```

These should be turned into `<` loops instead of `<=` loops if the result is off by one

Reply:
What you suggest looks much better! I assume there is no issue with invalidating the cache set-ways in the reverse order - actually that must be true as the DSB comes after the `do...while` loop in the original code.